api/test/service/upload.test.js

```diff
@@ -0,0 +1,367 @@
+import { testStore as test } from '../helpers/context.js'
+import { CreateTableCommand, QueryCommand } from '@aws-sdk/client-dynamodb'
+import { unmarshall } from '@aws-sdk/util-dynamodb'
+import * as Signer from '@ucanto/principal/ed25519'
+import * as UploadCapabilities from '@web3-storage/access/capabilities/upload'
+
+import getServiceDid from '../../authority.js'
+import { BATCH_MAX_SAFE_LIMIT } from '../../tables/upload.js'
+
+import { createDynamodDb } from '../utils.js'
+import { randomCAR } from '../helpers/random.js'
+import { getClientConnection, createSpace } from '../helpers/ucanto.js'
+
+test.beforeEach(async t => {
+  const region = 'us-west-2'
+  const tableName = 'upload'
+
+  // Dynamo DB
+  const {
+    client: dynamo,
+    endpoint: dbEndpoint
+  } = await createDynamodDb({ port: 8000, region })
+  await createDynamoUploadTable(dynamo)
+
+  t.context.dbEndpoint = dbEndpoint
+  t.context.dynamoClient = dynamo
+  t.context.tableName = tableName
+  t.context.region = region
+  t.context.serviceDid = await getServiceDid()
+})
+
+test('upload/add inserts into DB mapping between data CID and car CIDs', async (t) => {
+  const uploadService = t.context.serviceDid
+  const alice = await Signer.generate()
+  const { proof, spaceDid } = await createSpace(alice)
+  const connection = await getClientConnection(uploadService, t.context)
+
+  const car = await randomCAR(128)
+  const otherCar = await randomCAR(40)
+
+  // invoke a upload/add with proof
+  const root = car.roots[0]
+  const shards = [car.cid, otherCar.cid]
+
+  /** @type {import('../../service/types').UploadItemOutput[]} */
+  const uploadAdd = await UploadCapabilities.add.invoke({
+    issuer: alice,
+    audience: uploadService,
+    with: spaceDid,
+    nb: { root, shards },
+    proofs: [proof]
+    // @ts-expect-error ʅʕ•ᴥ•ʔʃ
+  }).execute(connection)
+
+  // @ts-expect-error error is added by ucanto if it fails
+  t.not(uploadAdd.error, true, uploadAdd.message)
+  t.is(uploadAdd.length, shards.length)
+
+  // Validate shards result
+  for (const shard of shards) {
+    const shardResult = uploadAdd.find((s) => s.carCID === shard.toString())
```

Comment:
minor: we should assert what we expect the order of shards to be here. not because we really care... _(tho i would want a test to fail if I changed that without meaning to)..._ but more because we should be very explicit about what we expect to see in the results.

if we sent `[shardA, shardA, shardB]` and the response contained `[shardA, shardB, shardB]`, this test would still pass. the length matches and we can find the item we want in it, but it is not correct.

Reply:
I will merge this as this needs discussion. 

If you consider `primaryIndex: { partitionKey: 'uploaderDID', sortKey: 'sk' },` the sortKey is `sk` which is `'dataCID#carCID'`. So there is no guarantee that the order is the order of insertion. But, I also don't see any requirement or need to keep an order. So, I consider that seeing we have length, and all are there is the expectation.